### PR TITLE
Migrate nteract play to TypeScript

### PR DESCRIPTION
**Pull request by captainsafia**
Types and such are fixed up.

Needs work to get styled-jsx working.

### CODE REVIEW

**Review by willingc — COMMENTED** [review]
Looking great so far :smile:

**Comment by captainsafia** [comment]
Migrated to styled-components to address the styled-jsx errors.

Now there is some stuff to fix with the tsconfig.

**Comment by captainsafia** [comment]
***meme-y kazoo sound***

This is ready for a review. 🎉

**Comment by rgbkrk** [comment]
Wahoo! Checking it out now!

File: applications/play/app-redux/actions.ts [hunk 1]
```diff
@@ -23,37 +25,42 @@ export const setSource = (payload: string) => ({
...
-  repo: string,
-  gitref: string
+  repo: string;
+  gitref: string;
 }) => ({
   type: actionTypes.SUBMIT_BINDER_FORM,
   payload
 });
-export const setCodeMirrorMode = (payload: string | Object) => ({
+export const setCodeMirrorMode = (payload: string | JSONObject) => ({
```
Comment (rgbkrk) [inline applications/play/app-redux/actions.ts:34]: 👍

File: packages/rx-binder/src/index.ts [hunk 2]
```diff
@@ -41,7 +41,7 @@ export const formBinderURL = ({
   return `${binderURL}/build/gh/${repo}/${ref}`;
 };
 
-interface IEventSource {
+export interface IEventSource {
```
Comment (rgbkrk) [inline packages/rx-binder/src/index.ts:44]: I didn't even realize we made this. Must have been created during the typescript migration and I was none-the-wiser. 😅

File: packages/messaging/src/types.ts [hunk 3]
```diff
@@ -7,6 +7,7 @@ export type MessageType =
   | "execute_request"
   | "inspect_request"
   | "kernel_info_request"
+  | "kernel_info_reply"
```
Comment (rgbkrk) [inline packages/messaging/src/types.ts:10]: 🎓

File: applications/play/next.config.js [hunk 4]
```diff
@@ -1,5 +1,4 @@
 const configurator = require("@nteract/webpack-configurator");
+const withTypescript = require("@zeit/next-typescript");
 
-module.exports = {
-  webpack: configurator.nextWebpack
-};
+module.exports = withTypescript({ webpack: configurator.nextWebpack });
```
Comment (rgbkrk) [inline applications/play/next.config.js:4]: Oh sweet, does this work super simply with this, even with no changes to our configurator? That's awesome. Makes the commuter transition more simple.

File: applications/play/package.json [hunk 5]
```diff
@@ -36,6 +36,13 @@
...
     "rx-jupyter": "^4.0.0-alpha.0",
     "rxjs": "^6.3.3",
+    "styled-components": "^4.1.2",
     "uuid": "^3.1.0"
+  },
+  "devDependencies": {
+    "@types/next": "^7.0.5",
+    "@types/react": "^16.7.13",
+    "@types/react-dom": "^16.0.11",
+    "typescript": "^3.2.2"
```
Comment (rgbkrk) [inline applications/play/package.json:46]: Does next-typescript implicitly include typescript as a regular dependency? I'm curious about what the prod build looks like.

File: applications/play/tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    "baseUrl": ".",
+    "paths": {
+      // Attempt to allow our namespace packages to get resolved
+      "@nteract/*": ["../../packages/*/src"],
+      "@mybinder/*": ["../../packages/*/src"],
+      "rx-jupyter": ["../../packages/rx-jupyter/src"],
+      "rx-binder": ["../../packages/rx-binder/src"],
+      "enchannel-zmq-backend": ["../../packages/enchannel-zmq-backend/src"],
+      "fs-observable": ["../../packages/fs-observable/src"]
+    }
```
Comment (rgbkrk) [inline applications/play/tsconfig.json:26]: Ooooh, this must have been all the magic to make this work!

Does the `composite: true` and `references` setup work here too?

**Review by rgbkrk — APPROVED** [review]
Super happy with this as the way we'll adapt other next.js apps! My only question inline is about what our chosen typescript config should be for these.

File: applications/play/package.json (hunk 5 above)
Comment (captainsafia) [reply applications/play/package.json:46]: It doesn't include TypeScript as a dependency. The [code](https://github.com/zeit/next-plugins/blob/master/packages/next-typescript/index.js) is basically some Webpack configs to get all their hot-reload stuff working. In the future, we might be able to integrate their logic with our webpack-configurator or just fallback to theirs.

File: applications/play/tsconfig.json (hunk 6 above)
Comment (captainsafia) [reply applications/play/tsconfig.json:26]: I actually included this from back when we were using the module remapper. Let me update it to use our composite set up and see if it still works.

**Comment by captainsafia** [comment]
OK. This is ready for review and merge. It works for developing locally but runs into the same commuter config issues when attempting to deploy in production. I've been researching how to get NextJS and have come across the following issues.

* [next-workspaces GH repo](https://github.com/panter/next-workspaces)
* [Active help discussion on monorepo support in Next with Yarn](https://spectrum.chat/next-js/general/getting-next-js-to-see-shared-code-in-yarn-monorepo~8e47a82a-26c1-48fb-804d-77dfc02a2286)
* [NextJS issue thread](https://github.com/zeit/next.js/issues/3638)

**Comment by rgbkrk** [comment]
Thanks for posting the links to resources while you investigate.

**Pull request merged**